aggregator/src/test/java/org/jenkinsci/plugins/workflow/cps/CpsThreadDumpTest.java

```diff
@@ -0,0 +1,108 @@
+    public JenkinsRule j = new JenkinsRule();
+    private WorkflowJob p;
+
+    @Before
+    public void setUp() throws Exception {
+        p = j.jenkins.createProject(WorkflowJob.class, "p");
+    }
+
+    @Test
+    public void threadDump() throws Exception {
+        p.setDefinition(new CpsFlowDefinition(StringUtils.join(asList(
+                "def foo() { bar() }",
+                "def bar() {",
+                "   semaphore 'x'",
+                "}",
+                "foo()"
+        ), "\n")));
+        WorkflowRun b = p.scheduleBuild2(0).waitForStart();
+        SemaphoreStep.waitForStart("x/1", b);
+        CpsFlowExecution e = (CpsFlowExecution) b.getExecution();
+        CpsThreadDump td = e.getThreadDump();
+        td.print(System.out);
+
+        {// verify that we got the right thread dump
+            List<ThreadInfo> threads = td.getThreads();
+            assertEquals(1, threads.size());
+            ThreadInfo t = threads.get(0);
+            assertThat(t.getHeadline(), containsString("Thread #0"));
+            assertStackTrace(t,
+                    "org.jenkinsci.plugins.workflow.test.steps.SemaphoreStep$Execution.<>(Native Method)",
+                    "WorkflowScript.bar(WorkflowScript:3)",
+                    "WorkflowScript.foo(WorkflowScript:1)",
+                    "WorkflowScript.run(WorkflowScript:5)");
+        }
+    }
+
+    @Test
+    public void threadDump2() throws Exception {
+        p.setDefinition(new CpsFlowDefinition(StringUtils.join(asList(
+                "def foo(x) { bar(x) }",// 1
+                "def bar(x) {",
+                "   semaphore x",       // 3
+                "}",
+                "def zot() {",
+                "  parallel(",          // 6
+                "    b1:{ foo('x') },",
+                "    b2:{ bar('y') });",
+                "}",
+                "zot()"                 // 10
+        ), "\n")));
+        WorkflowRun b = p.scheduleBuild2(0).waitForStart();
+        SemaphoreStep.waitForStart("x/1", b);
+        SemaphoreStep.waitForStart("y/1", b);
+
+        CpsFlowExecution e = (CpsFlowExecution) b.getExecution();
+        CpsThreadDump td = e.getThreadDump();
+        td.print(System.out);
+
+        assertStackTrace( td.getThreads().get(0),
+            "org.jenkinsci.plugins.workflow.test.steps.SemaphoreStep$Execution.<>(Native Method)",
+            "WorkflowScript.bar(WorkflowScript:3)",
+            "WorkflowScript.foo(WorkflowScript:1)",
+            "WorkflowScript.zot(WorkflowScript:7)",
+            "org.jenkinsci.plugins.workflow.cps.steps.ParallelStepExecution.<>(Native Method)",
+            "WorkflowScript.zot(WorkflowScript:6)",
+            "WorkflowScript.run(WorkflowScript:10)");
+
+        assertStackTrace( td.getThreads().get(1),
+            "org.jenkinsci.plugins.workflow.test.steps.SemaphoreStep$Execution.<>(Native Method)",
+            "WorkflowScript.bar(WorkflowScript:3)",
+            "WorkflowScript.zot(WorkflowScript:8)");
```

Comment:
A little odd that this does not show the rest of the stack that thread 0 shows, but OK—I guess that is considered redundant information?